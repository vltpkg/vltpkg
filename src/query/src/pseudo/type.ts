import { splitDepID } from '@vltpkg/dep-id/browser'
import type { ParserState } from '../types.ts'
import { asPostcssNodeWithChildren, asTagNode } from '../types.ts'
import { removeDanglingEdges, removeNode } from './helpers.ts'

/**
 * :type(str) Pseudo-Element will match only nodes that are of
 * the same type as the value used. The type is determined by the
 * first part of the dependency ID.
 */
export const type = async (state: ParserState) => {
  const type = asPostcssNodeWithChildren(state.current)
  const selector = asPostcssNodeWithChildren(type.nodes[0])
  const name = asTagNode(selector.nodes[0]).value
  for (const node of state.partial.nodes) {
    const nodeType = splitDepID(node.id)[0]
    if (nodeType !== name) {
      removeNode(state, node)
    }
  }
  removeDanglingEdges(state)
  return state
}
