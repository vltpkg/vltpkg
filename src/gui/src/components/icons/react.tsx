export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
