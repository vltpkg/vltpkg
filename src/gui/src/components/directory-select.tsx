import { useLayoutEffect, useRef, useState } from 'react'
import {
  Popover,
  PopoverTrigger,
  PopoverContent,
} from '@/components/ui/popover.tsx'
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
} from '@/components/ui/command.tsx'
import { Button } from '@/components/ui/button.tsx'
import { ChevronDown, Check } from 'lucide-react'
import { cn } from '@/lib/utils.ts'
import type { DashboardData } from '@/state/types.ts'

interface DirectorySelectProps {
  directory: string
  setDirectory: (directory: string) => void
  dashboard?: DashboardData
  acceptsGlobal?: boolean
  useDashboardProjectLocations?: boolean
  className?: string
}

export const DirectorySelect = ({
  directory,
  setDirectory,
  dashboard,
  acceptsGlobal = true,
  useDashboardProjectLocations = false,
  className,
}: DirectorySelectProps) => {
  const [displayPath, setDisplayPath] = useState<string>('')
  const mounted = useRef<boolean>(false)

  const directoryLocations =
    useDashboardProjectLocations ?
      dashboard?.dashboardProjectLocations
    : dashboard?.projects

  /**
   * The use a ref is to ensure that the initial display path is only set once.
   * This is because react will run useEffects twice in development mode.
   *
   * It is a useLayoutEffect to prevent layout shift on initial mount,
   * as it will run before initial paint.
   */
  useLayoutEffect(() => {
    const setInitialDisplayPath = () => {
      // query saved item case
      if (
        acceptsGlobal &&
        directory !== '' &&
        !useDashboardProjectLocations
      ) {
        setDisplayPath(
          directoryLocations?.find(
            location => location.path === directory,
          )?.readablePath ?? '',
        )
      }

      // query creation case
      else if (
        acceptsGlobal &&
        directory === '' &&
        !useDashboardProjectLocations
      ) {
        setDisplayPath('Global')
      }

      // creating a new project case
      else if (
        !acceptsGlobal &&
        useDashboardProjectLocations &&
        directory === ''
      ) {
        setDisplayPath('Select a directory')
      }
    }

    if (mounted.current) return
    mounted.current = true

    setInitialDisplayPath()
  }, [
    acceptsGlobal,
    directory,
    directoryLocations,
    useDashboardProjectLocations,
  ])

  return (
    <Popover>
      <PopoverTrigger asChild>
        <Button
          className={cn(
            'group flex h-[40px] w-full items-center justify-between border-[1px] border-muted bg-white text-muted-foreground shadow-none hover:bg-accent dark:bg-muted-foreground/5 dark:hover:bg-muted-foreground/10',
            className,
          )}>
          {displayPath}
          <ChevronDown className="text-foreground opacity-50 duration-300 group-data-[state=open]:-rotate-180" />
        </Button>
      </PopoverTrigger>
      <PopoverContent className="max-h-[--radix-popover-content-available-height] w-[--radix-popover-trigger-width] p-0">
        <Command defaultValue={acceptsGlobal ? '' : undefined}>
          <CommandInput placeholder="Directory" />
          <CommandList>
            <CommandEmpty>Directory not found.</CommandEmpty>
            <CommandGroup>
              {directoryLocations?.map(location =>
                location.readablePath === '~' && acceptsGlobal ?
                  <CommandItem
                    key={location.path}
                    value=""
                    className="hover:bg-accent"
                    onSelect={currentValue => {
                      setDirectory(currentValue === '' ? '' : '')
                      setDisplayPath('Global')
                    }}>
                    <Check
                      className={cn(
                        'mr-2 size-4',
                        directory === '' ? 'opacity-100' : (
                          'opacity-0'
                        ),
                      )}
                    />
                    Global
                  </CommandItem>
                : <CommandItem
                    key={location.path}
                    value={location.path}
                    onSelect={currentValue => {
                      setDirectory(
                        currentValue === directory ? '' : (
                          currentValue
                        ),
                      )
                      setDisplayPath(
                        acceptsGlobal && currentValue === directory ?
                          'Global'
                        : location.readablePath,
                      )
                    }}>
                    <Check
                      className={cn(
                        'mr-2 size-4',
                        directory === location.path ?
                          'opacity-100'
                        : 'opacity-0',
                      )}
                    />
                    {location.readablePath}
                  </CommandItem>,
              )}
            </CommandGroup>
          </CommandList>
        </Command>
      </PopoverContent>
    </Popover>
  )
}
