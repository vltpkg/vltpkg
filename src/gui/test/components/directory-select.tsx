import { vi, expect, afterEach, describe, it } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { useGraphStore as useStore } from '@/state/index.js'
import { DirectorySelect } from '@/components/directory-select.jsx'
import type { DashboardData } from '@/state/types.js'

vi.mock('@/components/ui/popover.jsx', () => ({
  Popover: 'gui-popover',
  PopoverTrigger: 'gui-popover-trigger',
  PopoverContent: 'gui-popover-content',
}))

vi.mock('@/components/ui/command.jsx', () => ({
  Command: 'gui-command',
  CommandEmpty: 'gui-command-empty',
  CommandGroup: 'gui-command-group',
  CommandInput: 'gui-command-input',
  CommandItem: 'gui-command-item',
  CommandList: 'gui-command-list',
}))

vi.mock('@/components/ui/button.jsx', () => ({
  Button: 'gui-button',
}))

vi.mock('lucide-react', () => ({
  ChevronDown: 'gui-chevron-down-icon',
  Check: 'gui-check-icon',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

describe('directory select', () => {
  it('renders correctly', () => {
    const mockSetDirectory = vi.fn()
    const mockDirectory = ''

    const mockDashboard = {
      buildVersion: '1.0.0',
      cwd: '/home/user',
      defaultAuthor: 'Ruy Adorno',
      dashboardProjectLocations: [
        { path: '/home/user', readablePath: '~' },
      ],
      projects: [
        {
          name: 'project-foo',
          readablePath: '~/project-foo',
          path: '/home/user/project-foo',
          manifest: { name: 'project-foo', version: '1.0.0' },
          tools: ['node', 'vlt'],
          mtime: 1730498483044,
        },
        {
          name: 'project-bar',
          readablePath: '~/project-foo',
          path: '/home/user/project-bar',
          manifest: { name: 'project-bar', version: '1.0.0' },
          tools: ['node', 'vlt'],
          mtime: 1730498491029,
        },
      ],
    } satisfies DashboardData

    const Container = () => {
      return (
        <DirectorySelect
          directory={mockDirectory}
          setDirectory={mockSetDirectory}
          dashboard={mockDashboard}
        />
      )
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toMatchSnapshot()
  })
})
