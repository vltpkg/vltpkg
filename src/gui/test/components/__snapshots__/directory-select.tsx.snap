// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`directory select > renders correctly 1`] = `

<gui-popover>
  <gui-popover-trigger aschild="true">
    <gui-button classname="group flex h-[40px] w-full items-center justify-between border-[1px] border-muted bg-white text-muted-foreground shadow-none hover:bg-accent dark:bg-muted-foreground/5 dark:hover:bg-muted-foreground/10">
      Global
      <gui-chevron-down-icon classname="text-foreground opacity-50 duration-300 group-data-[state=open]:-rotate-180">
      </gui-chevron-down-icon>
    </gui-button>
  </gui-popover-trigger>
  <gui-popover-content classname="max-h-[--radix-popover-content-available-height] w-[--radix-popover-trigger-width] p-0">
    <gui-command>
      <gui-command-input placeholder="Directory">
      </gui-command-input>
      <gui-command-list>
        <gui-command-empty>
          Directory not found.
        </gui-command-empty>
        <gui-command-group>
          <gui-command-item value="/home/user/project-foo">
            <gui-check-icon classname="mr-2 size-4 opacity-0">
            </gui-check-icon>
            ~/project-foo
          </gui-command-item>
          <gui-command-item value="/home/user/project-bar">
            <gui-check-icon classname="mr-2 size-4 opacity-0">
            </gui-check-icon>
            ~/project-foo
          </gui-command-item>
        </gui-command-group>
      </gui-command-list>
    </gui-command>
  </gui-popover-content>
</gui-popover>

`;
